src/librustc/traits/project.rs

```diff
@@ -1320,23 +1320,10 @@ fn assoc_ty_def<'cx, 'gcx, 'tcx>(
     let trait_def_id = selcx.tcx().impl_trait_ref(impl_def_id).unwrap().def_id;
     let trait_def = selcx.tcx().trait_def(trait_def_id);
 
-    if !trait_def.is_complete(selcx.tcx()) {
```

Comment:
So, the following gives me the error `overflow evaluating the requirement <u32 as Assoc>::Output` regardless of whether cycle detection is enabled:

```rust

#![feature(specialization)]

trait Assoc {
    type Output;
}

impl Assoc for u32 {
    type Output = u64;
}

impl Assoc for <u32 as Assoc>::Output {
    type Output = u64;
}

fn main() {
    let _x: <u64 as Assoc>::Output = 0;
}
```
So that's not really what we are looking for, I guess?